Enhance status for deploymentconfigs

kargakis opened the pull request:
Add various status fields for enhancing user experience. The fields are
also required for round-tripping upstream Deployments.

@ironcladlou @openshift/api-review  PTAL

Fixes https://github.com/openshift/origin/issues/5918
Part of https://trello.com/c/Mljldox7/643-5-deployments-downstream-support-for-upstream-features

@mfojtik @pweil- fyi

deads2k commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -137,7 +138,10 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.LatestVersion)
-	return nil
+	// Update deploymentConfig status; we probably need a separate call for status
```
If we're introducing the idea of a `status`, let's introduce a `status` at least internally.  Since `status` should be read-only, we can be compatible by duplicating existing fields in two locations.

Does the deployer pod update logical status?  If so, then we'll need to need to respect changes on it if the `status` stanza is missing.  If not and its only the baked in controller, then we should be good to ignore status updates on old fields.

kargakis replied:
The deployer pod updates the status on the replication controller, it never touches the config. I think that `status` would be internal only, @ironcladlou thoughts?

deads2k replied:
> The deployer pod updates the status on the replication controller, it never touches the config. I think that status would be internal only, @ironcladlou thoughts?

I'm fine with a deployer pod updating status.  I just wanted to make sure we handled it properly for backwards compatibility.  Sounds like we don't have to worry about it.

ironcladlou replied:
The strategies just scale RCs up and down and don't touch the config. They also have no reason to read the new field via public API or anything. So I don't think introducing status to the config should have an effect outside the controllers.

DirectXMan12 commented on pkg/deploy/api/types.go:
```diff
@@ -257,6 +257,8 @@ type DeploymentConfig struct {
 	// This could be based on a change made by the user or caused by an automatic trigger
 	Details *DeploymentDetails
+	// Replicas is the total number of pods targeted by this deploymentConfig.
+	Replicas int
```
This name could get confusion -- we now have `DeploymentConfig.Replicas` and `DeploymentConfig.Template.Replicas`, one of which is read-only, and one of which now scales the DeploymentConfig, and it's not particularly clear which is which based on the name.  Since we don't have a status object here, perhaps this should be `TotalReplicas`, `CurrentReplicas`, or something of the sort, just to avoid confusion.

ironcladlou replied:
I think I missed why we can't bring the `Status` field internal.

kargakis replied:
> I think I missed why we can't bring the Status field internal.

I am wondering about this. Our versioned dcs have status but not the internal one. Is there any reason why?

ironcladlou replied:
The internal looks pretty weird because when we refactored the model for v1 (from which upstream is derived), we didn't refactor the internal model to match due to time constraints.

kargakis replied:
OK, I think I want to change this here. Refactor the internal representation to use status and spec, add a status subresource and a separate call.

kargakis replied:
Status is now a thing for our internal object.

smarterclayton commented on pkg/deploy/api/v1/types.go:
```diff
@@ -262,6 +262,8 @@ type DeploymentConfigStatus struct {
 	// This could be based on a change made by the user or caused by an automatic trigger
 	Details *DeploymentDetails `json:"details,omitempty" description:"reasons for the last update to the config"`
+	// Replicas is the total number of pods targeted by this deploymentConfig.
+	Replicas int `json:"replicas,omitempty" description:"total number of pods targeted by this config"`
```
Is this upstream?

smarterclayton replied:
If so, you have API approval.

deads2k replied:
> Is this upstream?
> If so, you have API approval.

Upstream has a `status` object.

kargakis replied:
Which status object has `replicas` for total replicas.

kargakis commented:
Made this compile, can I have reviews especially for conversions (2nd commit) ? @deads2k @ironcladlou

kargakis commented:
Actually hack/update-generated-conversions generates all conversions needed so I guess we don't need manual conversions any more... There are also other new conversions added that I am not sure of ...

kargakis commented:
> There are also other new conversions added that I am not sure of ...

Hm, I think the rest are because of PodTemplateSpec...

deads2k commented on pkg/deploy/api/types.go:
```diff
@@ -245,27 +245,46 @@ const DeploymentCancelledAnnotationValue = "true"
+	Replicas int
+
+	// Selector is a label query over pods that should match the Replicas count.
+	Selector map[string]string
```
Is this new?  What was the existing field?

liggitt replied:
can a single selector represent the replicas? What if you are rolling between two deployments which resulted from two versions of the config, and have different selectors?

deads2k replied:
> can a single selector select all the replicas we care about? What if you are rolling between two deployments which resulted from two versions of the config, and have different selectors?

I asked @kargakis to split this into one pull for making spec/status with no other changes and then a separate pull to add the changes.  I *think* this is new, so that should help us find and see what this is doing.

kargakis replied:
The new fields were behind DeploymentTemplate previously.

deads2k replied:
> The new fields were behind DeploymentTemplate previously.

Ah, definitely interested in that separate pull then.  Checking congruence in one and then reasoning about changes in another..

Couldn't we have updated the `Selector` of the DC and orphaned existing pods for an old RC created by this DC?

kargakis commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -137,7 +138,10 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.Status.LatestVersion)
-	return nil
+	// Update deploymentConfig status; we probably need a separate call for status
```
Opened https://github.com/openshift/origin/issues/6465 for a separate status update call.

kargakis commented:
[test]

DirectXMan12 commented on pkg/deploy/registry/deployconfig/etcd/etcd.go:
```diff
@@ -152,48 +127,17 @@ func (r *ScaleREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object
 	}
+
+	scaleRet.Spec.Replicas = scale.Spec.Replicas
 	scaleRet.Status.Replicas = totalReplicas + (scale.Spec.Replicas - oldReplicas)
```
With @ironcladlou's change to how DeploymentConfig scaling works, I don't think this calculation is accurate any more. We should just keep `TotalReplicas == oldReplicas`, since the replica count on RCs controlled by the DC won't change until the DC controller updates them.

kargakis commented:
Flake in the deployment triggers integration test
```
I1222 13:09:46.141408   21420 controller.go:72] Waiting for first version of integration/config:0
I1222 13:09:46.145428   21420 controller.go:50] Created initial deployment for deploymentConfig "integration/config:0"
I1222 13:09:46.146497   21420 controller.go:61] Ignoring change for DeploymentConfig integration/config:1; no existing Deployment found
I1222 13:09:46.158297   21420 controller.go:74] Ignoring DeploymentConfig change for integration/config:1 (latestVersion=1); same as Deployment integration/config-1
--- FAIL: TestTriggers_configChange-4 (0.40s)
	deploy_trigger_test.go:319: Starting test openshift
	deploy_trigger_test.go:277: replicationControllers "config-1" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
FAIL
```
Hopefully will be fixed by https://github.com/openshift/origin/pull/6467

[test]

liggitt commented:
Nope, the same issue has happened with in-process etcd

kargakis commented:
https://github.com/openshift/origin/issues/6040 flake

kargakis commented:
```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.4.185:4001] twice [last error: Unexpected HTTP status code]) [0]
!!! Error in hack/../test/end-to-end/core.sh:162
	'oc delete pod cli-with-token' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:162 main(...)
Exiting with status 1
```

kargakis commented:
[test]

kargakis commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -247,6 +254,11 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
+	// TODO: Use a separate call for updating the status of a DC
+	// https://github.com/openshift/origin/issues/6465
+	config.Status.Replicas = kdeputil.GetReplicaCountForRCs(deployments)
+	_, err := c.osClient.DeploymentConfigs(config.Namespace).Update(config)
```
@ironcladlou won't this result in a conflict if an older client had scaled the DC (L220)?

ironcladlou replied:
I think you're right- shouldn't we be reassigning config local after the update on L220?

deads2k replied:
> I think you're right- shouldn't we be reassigning config local after the update on L220?

Not finding the conflict in a test, suggests we have a test missing.

It seems like this update is more properly a patch until we have separate status updating.

kargakis replied:
Would using another copy of the dc in the last update call not conflict?

kargakis replied:
> Not finding the conflict in a test, suggests we have a test missing.

We need to use older clients in our tests

kargakis replied:
Or in this specific case we could emulate an older client by manually updating the oldest rc for a dc

deads2k commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -137,7 +138,11 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.Status.LatestVersion)
-	return nil
+	// TODO: Use a separate call for updating the status of a DC
```
We'll want an issue for this.

kargakis replied:
https://github.com/openshift/origin/issues/6465

smarterclayton commented:
This API change is approved.

kargakis commented:
I would say that this is blocked on https://github.com/openshift/origin/issues/6465 (so we can avoid resource conflicts in case of old client updates) which in turn is blocked on https://github.com/openshift/origin/issues/6337 (the trickiest one; planning to start working on it soon)

stevekuznetsov commented on pkg/deploy/api/types.go:
```diff
@@ -314,14 +314,23 @@ type DeploymentConfigSpec struct {
+	// cofnig is out of sync.
 	LatestVersion int
-	// Details are the reasons for the update to this deployment config.
-	// This could be based on a change made by the user or caused by an automatic trigger
```
`s/cofnig/config/`

stevekuznetsov commented on pkg/deploy/api/types.go:
```diff
@@ -314,14 +314,23 @@ type DeploymentConfigSpec struct {
+	// change made by the user or caused by an automatic trigger.
 	Details *DeploymentDetails
-	// ObservedGeneration is the most recent generation observed by the controller.
+	// ObservedGeneration is the most recent generation observed by the deploymentconfig controller.
```
Two words?

stevekuznetsov replied:
Also I still think if we're not specially calling it a `DeploymentConfig` proper noun, since this is user facing doc, we should use "configuration" over "config".

stevekuznetsov commented on pkg/deploy/api/v1/types.go:
```diff
@@ -278,14 +278,23 @@ type DeploymentConfigSpec struct {
-	// LatestVersion is used to determine whether the current deployment associated with a DeploymentConfig
-	// is out of sync.
+	// LatestVersion is used to determine whether the current deployment associated with a deployment
+	// cofnig is out of sync.
```
Same tytrip as above

stevekuznetsov replied:
Typo, even.

stevekuznetsov commented:
I missed the discussion that spurred this PR five months ago... Why do we want this? This seems like useful information to show a user, but the aggregation of the info from RCs, etc, could be done client side. Why add this into these structs?

kargakis commented:
> I missed the discussion that spurred this PR five months ago... Why do we want this? This seems like useful information to show a user, but the aggregation of the info from RCs, etc, could be done client side. Why add this into these structs?

That's the job of a controller, to reconcile the spec and report on status. The dc controller already has a watch and knows all the required information for a given DC, why duplicate this work on clients? Also, we need the fields for parity with upstream Deployments.

kargakis commented:
@mfojtik @pweil- this is part of https://trello.com/c/Mljldox7/643-5-deployments-downstream-support-for-upstream-features

smarterclayton commented:
Each client should not have to recreate the world to know what is going
on.  A deployment config is a resource that abstracts many smaller resources

On May 12, 2016, at 6:52 PM, Michail Kargakis <notifications@github.com>
wrote:

I missed the discussion that spurred this PR five months ago... Why do we
want this? This seems like useful information to show a user, but the
aggregation of the info from RCs, etc, could be done client side. Why add
this into these structs?

That's the job of a controller, to reconcile the spec and report on status.
The dc controller already has a watch and knows all the required
information for a given DC, why duplicate this work on clients? Also, we
need the fields for parity with upstream Deployments.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6233#issuecomment-218908275>

stevekuznetsov commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
 	return nil
 }
+
+func calculateStatus(config deployapi.DeploymentConfig, deployments []kapi.ReplicationController) (*deployapi.DeploymentConfigStatus, error) {
```
The actual config is only used for the error messages, so I think this method should really just return the three counts and not consume the config at all, or update the given config (and consume it as a pointer).

kargakis replied:
Not really. It is used for latestVersion, status details, generation, and you are going to use it for minReadySeconds.

kargakis replied:
Also note that I am passing a copy and not a pointer.

stevekuznetsov commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
 		return transientError(fmt.Sprintf("cannot update the status for %q - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
+
+	if !reflect.DeepEqual(*newStatus, config.Status) {
```
Simply updating the status would mean we don't need this (costly?) reflection call, as well.

kargakis replied:
This action will reflect its cost only on the present process while a redundant call to the apiserver may lead in update conflicts on other clients.

stevekuznetsov replied:
What call to the api server do you mean?

kargakis replied:
`c.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config)`

The controller is a client to the apiserver as every other client.

stevekuznetsov replied:
I see. Fair enough, comments on the method remain !

kargakis replied:
This is WIP:) I will probably come back to it later today

stevekuznetsov commented on pkg/deploy/util/util.go:
```diff
@@ -233,6 +235,66 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
+
+// GetActualReplicaCountForDeployments returns the sum of actual replicas for the
+// given deployments.
+func GetActualReplicaCountForDeployments(deployments []api.ReplicationController) int {
```
"Actual" isn't making a meaningful distinction for me here - the above method returns the desired replicas, so perhaps "Actual" would mean more if that method was `GetDesiredReplicaCountForDeployments` our similar?

stevekuznetsov commented on pkg/deploy/util/util.go:
```diff
@@ -233,6 +235,66 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
+	allPods := map[string]api.Pod{}
+	for _, deployment := range deployments {
+		options := api.ListOptions{LabelSelector: labels.Set(deployment.Spec.Selector).AsSelector()}
+		podList, err := c.Pods(deployment.ObjectMeta.Namespace).List(options)
```
Won't this alias pods with the same name under different namespaces?

kargakis replied:
This helper is called for calculating the available pods for a deploymentconfig so all deployments and pods will be at the same namespace. I will add a comment clarifying it.

stevekuznetsov commented on pkg/deploy/controller/deploymentconfig/controller.go:
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
+
+func calculateStatus(config deployapi.DeploymentConfig, deployments []kapi.ReplicationController) (*deployapi.DeploymentConfigStatus, error) {
+	// TODO: Implement MinReadySeconds for deploymentconfigs: https://github.com/openshift/origin/issues/7114
+	available, err := deployutil.GetAvailablePodsForDeployments(c.kubeClient, deployments, 0)
```
Name the int literal here for clarity

stevekuznetsov replied:
Would clarify the TODO as well

kargakis replied:
ok

kargakis commented:
@openshift/api-review can I get a review here?

kargakis commented:
@smarterclayton I think these new fields are at the mercy of the resync interval of the deploymentconfig controller.

kargakis commented:
I have identified two spots where status updates need to happen: one is after a deployment finishes (we could also have status updates from the deployer pod while it is scaling up/down) and one after scaling. The first case is covered by https://github.com/openshift/origin/pull/6233/commits/ae8c60b319a2ecddce183e9f4cefe1edf8eee361, but in the second case I am not sure what else is left except the resync interval. By the time etcd triggers a watch for the controller for a .spec.replicas change, most of the pods will not be ready, the dc will be reconciled and its status will be actually updated in the next resync.

smarterclayton commented:
API approved.

kargakis commented:
@mfojtik a review here can relax you from the rebase.

mfojtik commented on pkg/deploy/api/helpers.go:
```diff
@@ -32,8 +33,19 @@ func ScaleFromConfig(dc *DeploymentConfig) *extensions.Scale {
 			CreationTimestamp: dc.CreationTimestamp,
 		},
+		Spec: extensions.ScaleSpec{
+			Replicas: int(dc.Spec.Replicas),
```
this will be changed to int32() in rebase I guess.

kargakis replied:
Yeap. Waiting for it;)

mfojtik commented on pkg/deploy/controller/deploymentconfig/factory.go:
```diff
@@ -40,7 +40,7 @@ func (factory *DeploymentConfigControllerFactory) Create() controller.RunnableCo
 	}
 	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 2*time.Minute).Run()
+	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 30*time.Second).Run()
```
why this change?

kargakis replied:
I still need @smarterclayton to approve this. It's the only issue with this PR. The upstream controller has the same sync interval so status for upstream Deployments will usually be updated faster than having it set to 2 minutes. See https://github.com/openshift/origin/pull/6233#issuecomment-219423116

smarterclayton replied:
Hrm - I guess the next rebase (after this) we'll pull the list vs sync separation.  I'm concerned a bit that this will regress us a bit in performance, and whether we'll have to do a lot of work to pick that up.

@rrati do your changes to resync/relist depend on Informers, or can they work for any reflector?

rrati replied:
That really depends on what is needed.:) The reflector itself will not exit on a resync notification any more.  It will break the watch, call Resync on the store object the reflector was configured to use, then initiate the watch.

However, the DeltaFIFO was the only store implementation that had access to the local cache and a notification means to allow the resync to cause the local cache to be inspected.  Other storage types will do nothing on a resync or just ensure that all items in the fifo are in the processing queue for the workers to process.

So, while the reflector will not exit on a resync call any more, I would say only the informer (or reflectors configured similarly to the informer) will actually cause an evaluation of the local cache.  Whether or not that is a big deal depends on what the use case is.

smarterclayton replied:
Ok, so let's *not* change this for now, leave a TODO, and then someone
needs to own ensuring that the Origin controllers are taking advantage of
resync/relist correctly when it lands.

On Wed, May 18, 2016 at 1:58 PM, Rob Rati <notifications@github.com> wrote:

> In pkg/deploy/controller/deploymentconfig/factory.go
> <https://github.com/openshift/origin/pull/6233#discussion_r63751350>:
>
> > @@ -40,7 +40,7 @@ func (factory *DeploymentConfigControllerFactory) Create() controller.RunnableCo
> >  		},
> >  	}
> >  	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
> > -	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 2*time.Minute).Run()
> > +	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 30*time.Second).Run()
>
> That really depends on what is needed.:) The reflector itself will not
> exit on a resync notification any more. It will break the watch, call
> Resync on the store object the reflector was configured to use, then
> initiate the watch.
>
> However, the DeltaFIFO was the only store implementation that had access
> to the local cache and a notification means to allow the resync to cause
> the local cache to be inspected. Other storage types will do nothing on a
> resync or just ensure that all items in the fifo are in the processing
> queue for the workers to process.
>
> So, while the reflector will not exit on a resync call any more, I would
> say only the informer (or reflectors configured similarly to the informer)
> will actually cause an evaluation of the local cache. Whether or not that
> is a big deal depends on what the use case is.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6233/files/23f2869e9501ebeb135cad392256f527d16b67a0#r63751350>
>

kargakis replied:
Ok, got this working on top of https://github.com/openshift/origin/pull/9002 without the need to reduce the interval. Now all I need is the rebase to land so we can start using shared caches @mfojtik :)

mfojtik commented:
@kargakis LGTM, nice job! but it does not really relaxed me :)

ironcladlou commented on pkg/deploy/api/v1/types.go:
```diff
@@ -278,14 +278,23 @@ type DeploymentConfigSpec struct {
+	Replicas int32 `json:"replicas,omitempty"`
+	// UpdatedReplicas is the total number of non-terminated pods targeted by this deployment config
+	// that have the desired template spec.
+	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
```
Is it useful/important to know the updated available count?

kargakis replied:
Hm, we should discuss it upstream. For now, I think we are ok with AvailableReplicas.

smarterclayton replied:
Which of these are not upstream?

kargakis replied:
Every field is upstream already

On Thu, Jun 23, 2016 at 4:36 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/deploy/api/v1/types.go
> <https://github.com/openshift/origin/pull/6233#discussion_r68246188>:
>
> >  	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
> > +	// Replicas is the total number of pods targeted by this deployment config.
> > +	Replicas int32 `json:"replicas,omitempty"`
> > +	// UpdatedReplicas is the total number of non-terminated pods targeted by this deployment config
> > +	// that have the desired template spec.
> > +	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
>
> Which of these are not upstream?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/6233/files/23f2869e9501ebeb135cad392256f527d16b67a0#r68246188>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADuFf6JMAY9Jc3tBwTd-gsUEqEY4eKw5ks5qOpnogaJpZM4GxCXA>
> .
>

kargakis replied:
Except the one @ironcladlou proposes which is not in both

smarterclayton replied:
Ok, wanted to be clear.

ironcladlou commented:
Overall design LGTM, needs tests now.

kargakis commented:
The upstream controller [identifies updates on children (replicasets) and requeues the parent (deployment)](https://github.com/kubernetes/kubernetes/blob/62a8394eb491e82e5445e36542907608bf7b95f3/pkg/controller/deployment/deployment_controller.go#L242-L265). Doing that would most probably solve the issue with the resync we have here. We need to transition to using caches in the deployment controllers (right now we don't really use them).

kargakis commented:
> The upstream controller identifies updates on children (replicasets) and requeues the parent (deployment). Doing that would most probably solve the issue with the resync we have here. We need to transition to using caches in the deployment controllers (right now we don't really use them).

@smarterclayton this means that we need to depart from the RetryController framework and start using informers. Shared caches will also be another thing. cc: @deads2k

smarterclayton commented:
Yes.

On Mon, May 23, 2016 at 8:23 AM, Michail Kargakis <notifications@github.com>
wrote:

> The upstream controller identifies updates on children (replicasets) and
> requeues the parent (deployment). Doing that would most probably solve the
> issue with the resync we have here. We need to transition to using caches
> in the deployment controllers (right now we don't really use them).
>
> @smarterclayton <https://github.com/smarterclayton> this means that we
> need to depart from the RetryController framework and start using
> informers. Shared caches will also be another thing. cc: @deads2k
> <https://github.com/deads2k>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6233#issuecomment-220965287>
>

kargakis commented:
Rebased on top of the latest controller changes, shared a pod cache, added an extended test. This will need another look @mfojtik @ironcladlou @smarterclayton @deads2k

kargakis commented:
[test]

mfojtik commented:
LGTM

kargakis commented:
[merge]

kargakis commented:
This will also enable `oc rollout status`.
